[incubator/fluentd] Generalize chart to work with different backends

goruha opened the pull request:
## What
* Made working with environment variables flexible
* Made Fluentd be daemon set
* Remove health check
* Parametrise config mount dir
* Mount host logs into the container

## Why
* Make chart be flexible enough to work with different fluentd containers (ex. fluentd-cloudwatch)
* To store secrets private
* DeamonSet allow collecting logs from all nodes
* Not all fluentd containers support health checks
* Different fluentd containers expect configs in a different paths
* Allow collecting host logs with fluentd `in_tail`

goruha committed:
Made working with envvars flexible

goruha commented:
/assign @prydonius

goruha committed:
Redo deamonset

prydonius commented:
/assign @rendhalver

rendhalver commented on incubator/fluentd/templates/deamonset.yaml:
```diff
@@ -1,5 +1,5 @@
-apiVersion: apps/v1beta1
-kind: Deployment
+apiVersion: extensions/v1beta1
+kind: DaemonSet
```
Why are you switching this to a DaemonSet?
There is an fluentd-elasticsearch chart that does exactly this.

goruha replied:
that's true.
I'd like to make chart general enough for the different use of fluentd.
there are already `fluentd-elasticsearch`, `fluentd-clouwatch`.
I use fluentd chart with datadog.
It seems to be a mistake to create a chart for each type of backend. 
What do you think about?

rendhalver replied:
I like the idea but if we are going to replace all three charts with one it needs to be flexible and configurable enough to do what each of those chart do already without reducing the functionality the current Charts.
I wrote this one to replace logstash in my Elastic-Stack cluster so it didn't need to be a DaemonSet or need access to node logs.

mattfarina replied:
Two things...

1. If you're going to use something from the workloads api (e.g., DaemonSet) use it from the apps API group. Extensions should not be used anymore.
1. I like the idea of a fluentd chart that can have handle being installed with various backends. Yet, having different ones for each backend is ok, too. But, we should pick one way and stay consistent with it. If there is one it should cleanly handle all the cases and needs for all of them.

rendhalver replied:
If the DaemonSet can be still used as part of a Service then that works for me.
We use this chart to route logs from devices outside our cluster to an ES cluster in Kube so I would like to keep that functionality.

osterman replied:
(nitpick: `deamonset.yaml` misspelled)

osterman replied:
> If the DaemonSet can be still used as part of a Service then that works for me

A `Service` works by selecting `Pods` according to a `selector` so it's compatible with `DaemonSets` which create pods. A `Service` doesn't care (or know) how the `Pods` were created.

rendhalver replied:
OK awesome. Thank you.

rendhalver commented on incubator/fluentd/templates/deamonset.yaml:
```diff
@@ -49,23 +62,29 @@ spec:
-          initialDelaySeconds: 5
-          timeoutSeconds: 1
         volumeMounts:
+        - name: varlog
```
See above. You are basically turning this into the fluentd-elasticsearch chart.
If you are trying to merge these two please make these mounts optional.
There are uses for fluentd that don't involve container logs.

goruha replied:
@rendhalver would it be problem if we will mount logs inside container that do not require them?
I think we need mount logs always and simplify chart install

rendhalver replied:
Why?
It's two extra lines to exclude the mounts.
Not every implementation needs them and not all pods should have access to them.

osterman commented on incubator/fluentd/values.yaml:
```diff
@@ -9,11 +9,15 @@ image:
-  buffer_chunk_limit: 2M
-  buffer_queue_limit: 8
+env:
+  open:
```
`configMap` would be more clear

goruha replied:
@osterman `env` shows that we config container with environment variables

rendhalver replied:
I agree with @osterman.
Config maps are clearer.

goruha replied:
@osterman @rendhalver can you provide the example of the structure that would you prefer?
I can not get what is the problem

monotek commented:
Deamonsets are not really updateable as most of the fields are immutable. So using a damon set should be optional.

goruha commented:
@monotek we will use https://kubernetes.io/docs/tasks/manage-daemon/update-daemon-set/

goruha committed:
Address PR comments

goruha committed:
Address PR comments

goruha committed:
Address PR comments

goruha commented:
@osterman @mattfarina  @monotek @rendhalver 
I addressed your comments and provide readme with examples how to use this chart with different backends.
Hope you would find chart flexible enough and ready to release

monotek commented:
@goruha
I still think daemonset or deployment should be optional. 
Maybe one don't wants to use it for kubernetes / docker logs but for sending app logs to the fluentd forward plugin on port 24224. 
In this case 1 pod created via a deployment would be enough.

rendhalver commented:
I agree with @monotek here too.
We should have the option of running as a DaemonSet or a Deployment.

rendhalver commented on incubator/fluentd/templates/deamonset.yaml:
```diff
@@ -29,15 +34,26 @@ spec:
       - name: {{ .Chart.Name }}
         image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
         imagePullPolicy: {{ .Values.image.pullPolicy }}
+        securityContext:
```
This should be dependent on whether the logs are mounted.
If the pod doersn't need access to the logs if shouldn't have root access.

rendhalver commented on incubator/fluentd/templates/deamonset.yaml:
```diff
@@ -29,15 +34,26 @@ spec:
-            value: {{ .Values.output.buffer_chunk_limit | quote }}
-          - name: OUTPUT_BUFFER_QUEUE_LIMIT
-            value: {{ .Values.output.buffer_queue_limit | quote }}
+{{- range $name, $value := .Values.env.open }}
```
Why not just have one less level here and put the env vars in env.
Are you planning on putting anything else in this array?

goruha replied:
@rendhalver can you provide example what structure do you prefer?

rendhalver replied:
env: instead of env.open:
one less level to the data structure

rendhalver commented on incubator/fluentd/templates/deamonset.yaml:
```diff
@@ -49,23 +65,29 @@ spec:
           - name: http-input
             containerPort: 9880
             protocol: TCP
-        livenessProbe:
```
Why remove this?
We have health checks for a reason.
Do we not want to know if the pods are functioning correctly before we route traffic to them?

goruha replied:
Because liveness probe required config to provide the endpoint for the check. 
Not all containers support such configuration.
Instead of that, I disabled `supervisor` mode, so pod will failed if something goes wrong.
https://github.com/kubernetes/charts/pull/5357/files#diff-6ec61ff69620bf74339bd25610e91021R20

rendhalver replied:
I don't see how that achieves the same ends.

rendhalver commented on incubator/fluentd/values.yaml:
```diff
@@ -49,13 +30,6 @@ configMaps:
       @type null
     </match>
 
-    # Used for health checking
```
See above.

goruha replied:
See above

monotek commented:
Imho we should also provide 2 configmaps for optional deployment / daemonset.

If deployment is used I'm pretty sure most of the current config is unwanted.
